lib/evaluate/intrinsics.cc

```diff
@@ -1165,6 +1201,16 @@ std::optional<SpecificCall> IntrinsicInterface::Match(
         }
       }
       break;
+    case KindCode::operand:
+      CHECK(operandArg != nullptr);
+      if (std::optional<DynamicType> aType{operandArg->GetType()}) {
+        if (result.categorySet.test(aType->category())) {
+          resultType = *aType;
+        } else {
+          resultType = DynamicType{*category, aType->kind()};
```

Comment:
It's checked in the constructor for `DynamicType`.

I'll check whether the code can be simplified.  I copied it from the handling of `same` results, but maybe it can be simplified.